### ISSUE TITLE
add (with recursive: true) returns improperly-formatted JSON

### ISSUE THREAD

**Issue by tinybike** [ipfs/js-ipfs-api#82]
The `add` method (with `recursive: true`) isn't returning properly-formatted JSON for me.  Instead I get a string that almost looks like JSON, but isn't actually, so to parse it I have to do some string mangling.  Example:
```javascript
var ipfs = require("ipfs-api")("localhost", "5001");
ipfs.add("./ipfs", {recursive: true}, function (err, res) {
            if (err || !res) return console.error(err);
            console.log(res);
});
```
This prints out:
```javascript
{
  "Name": "funcrusher.json",
  "Hash": "QmUSpeNRKrYHRG55WvYTi5u3VVHaP3uHtH1mDKCB9Qa9AJ"
}{
  "Name": "ipfs",
  "Hash": "QmYuDw2SxfWkFasmmenzsdKSV7QVHAEuevAuCjcDh8kEYt"
}{
  "Name": "",
  "Hash": "QmWike1KhJRzY4EMtreidZDgq1EhYpLY444pfVJxxD3wSp"
}
```
(Note the lack of commas between the objects, and lack of enclosing array brackets.)

**Comment by travisperson** [comment]
This might be an issue with the daemon. It would appear that it may not be properly chunking the response coming back. If I had to guess the `x-chunked-output` header is missing.

**Comment by whyrusleeping** [comment]
many of our commands return a stream of json objects, not a json array. This is so that consumers of the api can act on the output before receiving the entirety of the output. Most json decoders (all of them that i've used) are able to handle this pretty well.

The one issue that we have though, is that we should be putting newlines between the objects to be true 'ndjson'.

**Comment by whyrusleeping** [comment]
the newline part should be fixed here: https://github.com/ipfs/go-ipfs/pull/1885

**Comment by travisperson** [comment]
This also may be causing issues: ipfs/go-ipfs#1882

**Comment by diasdavid** [comment]
@tinybike IPFS HTTP API started returning ndjson on 0.3.8, while in 0.3.7 returned a JSON array. This module is expecting a JSON array and trying to perform a parse on it, but since it is not valid JSON, it returns it as a string. We are working to upgrade it, but prioritised tests first https://github.com/ipfs/node-ipfs-api/pull/81, once tests are ready (tested against 0.3.7) we will add a proper ndjson parser. The focus on tests first is exactly to avoid this kind of bugs with API changes in the future. Sorry for the trouble, confusion.

**Comment by diasdavid** [comment]
solved with latest release and IPFS 0.3.9 :)
